src/hotspot/share/gc/parallel/psParallelCompact.cpp

```diff
@@ -295,110 +304,109 @@ ParallelCompactData::summarize_dense_prefix(HeapWord* beg, HeapWord* end)
+    size_t preceding_live_words;
+    if (is_region_aligned(overflowing_obj)) {
+      preceding_live_words = 0;
+    } else {
+      // Words accounted by the overflowing object on the split region
+      size_t overflowing_size = pointer_delta(region_align_up(overflowing_obj), overflowing_obj);
+      preceding_live_words = region(split_region)->data_size() - overflowing_size;
     }
 
-    // Set split_destination and partial_obj_size to reflect the split region.
-    split_destination = sr->destination();
-    partial_obj_size = sr->partial_obj_size();
-  }
+    split_info.record(split_region, overflowing_obj, preceding_live_words);
+
+    HeapWord* src_region_start = region_to_addr(src_region);
+    HeapWord* new_top = destination - pointer_delta(src_region_start, overflowing_obj);
+
+    // If the overflowing obj was relocated to its original destination,
+    // those destination regions would have their source_region set. Now that
+    // this overflowing obj is relocated somewhere else, reset the
+    // source_region.
+    {
+      size_t range_start = addr_to_region_idx(region_align_up(new_top));
+      size_t range_end = addr_to_region_idx(region_align_up(destination));
+      for (size_t i = range_start; i < range_end; ++i) {
+        region(i)->set_source_region(0);
+      }
+    }
+
+    // Update new top of target space
+    *target_next = new_top;
 
-  // The split is recorded only if a partial object extends onto the region.
-  if (partial_obj_size != 0) {
-    _region_data[split_region].set_partial_obj_size(0);
-    split_info.record(split_region, partial_obj_size, split_destination);
+    return overflowing_obj;
   }
 
-  // Setup the continuation addresses.
-  *target_next = split_destination + partial_obj_size;
-  HeapWord* const source_next = region_to_addr(split_region) + partial_obj_size;
+  // Obj-iteration to locate the overflowing obj
+  HeapWord* region_start = region_to_addr(src_region);
+  HeapWord* region_end = region_start + RegionSize;
+  HeapWord* cur_addr = region_start + partial_obj_size;
+  size_t live_words = partial_obj_size;
 
-  if (log_develop_is_enabled(Trace, gc, compaction)) {
-    const char * split_type = partial_obj_size == 0 ? "easy" : "hard";
-    log_develop_trace(gc, compaction)("%s split:  src=" PTR_FORMAT " src_c=" SIZE_FORMAT " pos=" SIZE_FORMAT,
-                                      split_type, p2i(source_next), split_region, partial_obj_size);
-    log_develop_trace(gc, compaction)("%s split:  dst=" PTR_FORMAT " dst_c=" SIZE_FORMAT " tn=" PTR_FORMAT,
-                                      split_type, p2i(split_destination),
-                                      addr_to_region_idx(split_destination),
-                                      p2i(*target_next));
+  while (true) {
+    assert(cur_addr < region_end, "inv");
+    cur_addr = PSParallelCompact::mark_bitmap()->find_obj_beg(cur_addr, region_end);
+    if (cur_addr >= region_end) {
```

Comment:
Shouldn't this be changed to an assert?

Given the:
`// There must be an overflowing obj in this region`
`ShouldNotReachHere();`

if the `break` is executed.